cpp/include/rapidsmpf/buffer/buffer.hpp

```diff
@@ -136,31 +186,35 @@ class Buffer {
     }
 
     /**
-     * @brief Override the event for the buffer.
+     * @brief Check whether the buffer's most recent write has completed.
      *
-     * @note Use this if you want the buffer to sync with an event happening after the
-     * original event. Need to be used with care when dealing with multiple streams.
+     * Returns whether the CUDA event that tracks the most recent write into this
+     * buffer has been signaled.
      *
-     * @param event The event to set.
-     */
-    void override_event(std::shared_ptr<CudaEvent> event) {
-        event_ = std::move(event);
-    }
-
-    /**
-     * @brief Check if the device memory operation has completed.
+     * Use this to guard *non-stream-ordered* consumer-APIs that do not accept a CUDA
+     * stream (e.g., MPI sends/receives, host-side reads).
      *
-     * @return true if the device memory operation has completed or no device
-     * memory operation was performed, false if it is still in progress.
-     */
-    [[nodiscard]] bool is_ready() const;
-
-    /**
-     * @brief Wait for the device memory operation to complete.
+     * @note This is a non-blocking, point-in-time status check and is subject to TOCTOU
+     * races: another thread may enqueue additional writes after this returns `true`.
+     * Ensure no further writes are enqueued, or establish stronger synchronization (e.g.,
+     * synchronize the buffer's stream) before using the buffer.
+     *
+     * @return `true` if the last recorded write event has completed; `false` otherwise.
      *
-     * @throws rapidsmpf::cuda_error if event wait fails (if set).
+     * @code{.cpp}
+     * // Example: send the buffer via MPI (non-stream-ordered).
+     * if (buffer.is_latest_write_done()) {
+     *   MPI_Isend(buffer.data(), buffer.size(), MPI_BYTE, dst, tag, comm, &req);
+     * } else {
+     *   // Ensure completion before handing to MPI.
+     *   cudaStreamSynchronize(buffer.stream());
```

Comment:
To facilitate this usage, and to avoid oversynchronisation, we should introduce `buffer.wait_for_writes()` that does `cudaEventSynchronize` on the `latest_write_event_`